clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/AbstractEcsCachingAgent.java

```diff
@@ -0,0 +1,166 @@
+  /**
+   * Fetches items from the ECS service.
+   * @param ecs The AmazonECS client that will be used to make the queries.
+   * @param providerCache A ProviderCache that is used to access already existing cache.
+   * @return A list of generic type objects.
+   */
+  protected abstract List<T> getItems(AmazonECS ecs, ProviderCache providerCache);
+
+  /**
+   * Generates a map of CacheData collections associated to a key namespace from a given collection of generic type objects.
+   * @param cacheableItems A collection of generic type objects.
+   * @return A map of CacheData collections belonging to a key namespace.
+   */
+  protected abstract Map<String, Collection<CacheData>> generateFreshData(Collection<T> cacheableItems);
+
+  @Override
+  public String getProviderName() {
+    return EcsProvider.NAME;
+  }
+
+  @Override
+  public CacheResult loadData(ProviderCache providerCache) {
+    String authoritativeKeyName = getAuthoritativeKeyName();
+
+    AmazonECS ecs = amazonClientProvider.getAmazonEcs(accountName, awsCredentialsProvider, region);
+    List<T> items = getItems(ecs, providerCache);
+    return buildCacheResult(authoritativeKeyName, items, providerCache);
+  }
+
+  /**
+   * Provides a set of ECS cluster ARNs.
+   * Either uses the cache, or queries the ECS service.
+   * @param ecs The AmazonECS client to use for querying.
+   * @param providerCache The ProviderCache to retrieve clusters from.
+   * @return A set of ECS cluster ARNs.
+   */
+  Set<String> getClusters(AmazonECS ecs, ProviderCache providerCache) {
+    Set<String> clusters = providerCache.getAll(ECS_CLUSTERS.toString()).stream()
+      .map(cacheData -> (String) cacheData.getAttributes().get("clusterArn"))
+      .collect(Collectors.toSet());
+
+    if (clusters == null || clusters.isEmpty()) {
+      clusters = new HashSet<>();
+      String nextToken = null;
+      do {
+        ListClustersRequest listClustersRequest = new ListClustersRequest();
+        if (nextToken != null) {
+          listClustersRequest.setNextToken(nextToken);
+        }
+        ListClustersResult listClustersResult = ecs.listClusters(listClustersRequest);
+        clusters.addAll(listClustersResult.getClusterArns());
+
+        nextToken = listClustersResult.getNextToken();
+      } while (nextToken != null && nextToken.length() != 0);
+    }
+
+    return clusters;
+  }
+
+  /**
+   * Provides the key namespace that the caching agent is authoritative of.
+   * Currently only supports the caching agent being authoritative over one key namespace.
+   * @return Key namespace.
+   */
+  String getAuthoritativeKeyName() {
+    Collection<AgentDataType> authoritativeNamespaces = getProvidedDataTypes().stream()
+      .filter(agentDataType -> agentDataType.getAuthority().equals(AUTHORITATIVE))
+      .collect(Collectors.toSet());
+
+    if (authoritativeNamespaces.size() != 1) {
+      throw new RuntimeException("AbstractEcsCachingAgent supports only one authoritative key namespace.");
```

Comment:
Perhaps including the actual size of the collection in the exception would be valuable? `AbstractEcsCachingAgent supports only one authoritative key namespace (%d given).` sorta thing

Reply:
Done.